### PR TITLE
Migrate to Webpack 3.x

### LINKED ISSUE #367: Upgrade to Webpack 3
And remove/replace `webpack-blocks` in the process? 
  _We _currently_ have to since it only supports `1.x` and `2.x`_

### PR DESCRIPTION
**Pull request by lukeed**
- Drops the dependencies on `@webpack-blocks/*` which was explicitly requiring `webpack@2.x` and didn't look like an update was on the horizon.

- Includes `webpack.ModuleConcatenationPlugin` on client builds.

- Installs style-type loaders that `@webpack-blocks` installed for us previously

- Merges `webpack` configuration values via `webpack-merge`

Thanks to @hassanbazzi who did most of the config rewrite in #150. This PR is largely based off his work 👏 This exists as a separate PR because some of Hassan's original changes were lost over time during merge-commits & this was easier (imo) than trying to go back and fix them.

---

Closes #367, #373, #150 _(outdated)_

### CODE REVIEW

**Comment by lukeed** [comment]
@reznord Lol, it's not possible 😜

File: src/lib/webpack/prerender.js [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 import { SourceMapConsumer } from 'source-map';
-import chalk from 'chalk';
 
 export default function prerender(env, params) {
 	params = params || {};
 
-	let entry = resolve(env.dest, './ssr-build/ssr-bundle.js'),
-		url = params.url || '/';
+	let entry = resolve(env.dest, './ssr-build/ssr-bundle.js');
+	let url = params.url || '/';
```
Comment (hassanbazzi) [inline src/lib/webpack/prerender.js]: is `let` neccesary here? Can this be a `const`?

Comment (lukeed) [reply src/lib/webpack/prerender.js]: @developit prefers `let` for most things. I broke the rule & used `const` a few times to be emphatic.

File: src/lib/webpack/render-html-plugin.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		preload: config.preload===true,
+		title: title || config.title || config.manifest.name || config.manifest.short_name || (config.pkg.name || '').replace(/^@[a-z]\//, '') || 'Preact App',
+		excludeAssets: [/(bundle|polyfills)(\..*)?\.js$/],
+		config,
+		ssr(params) {
+			return config.prerender ? prerender({ cwd, dest, src }, { ...params, url }) : '';
+		}
+	});
+
+	const pages = readJson(resolve(cwd, config.prerenderUrls || '')) || [{ url: '/' }];
```
Comment (hassanbazzi) [inline src/lib/webpack/render-html-plugin.js]: Should't this be using `params.url` for the fallback?

File: src/lib/webpack/webpack-client-config.js [hunk 3]
```diff
@@ -1,132 +1,167 @@
...
-		setOutput({
+	return {
+		entry: {
+			bundle: resolve(__dirname, './../entry'),
+			polyfills: resolve(__dirname, './polyfills')
+		},
+
+		output: {
 			path: env.dest,
 			publicPath: '/',
```
Comment (hassanbazzi) [inline src/lib/webpack/webpack-client-config.js]: Can we use params.url here?

File: src/lib/webpack/render-html-plugin.js (hunk 2 above)
Comment (lukeed) [reply src/lib/webpack/render-html-plugin.js]: For `pages`? The `params` is only within the `ssr` function. I think the `/` route is a safe default. Plus, it's just copied over from what was there already.

Comment (hassanbazzi) [reply src/lib/webpack/render-html-plugin.js]: Yea maybe we should do it in a separate PR. Basically, the whole "subdirectory" stuff is something we should fix. But yea, let's keep it separate.

**Comment by hassanbazzi** [comment]
I won't be able to test until later tonight if that's ok :)

**Comment by lukeed** [comment]
All good~!

**Comment by reznord** [comment]
@lukeed every thing works pixel perfect 😄

**Comment by reznord** [comment]
Merging this.

**Pull request merged**